drivers/LargeBlockSR.py

```diff
@@ -0,0 +1,203 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import SR
+from SR import deviceCheck
+import SRCommand
+import EXTSR
+import util
+import xs_errors
+import os
+import re
+import lvutil
+
+CAPABILITIES = ["SR_PROBE", "SR_UPDATE", "SR_SUPPORTS_LOCAL_CACHING",
+                "VDI_CREATE", "VDI_DELETE", "VDI_ATTACH", "VDI_DETACH",
+                "VDI_UPDATE", "VDI_CLONE", "VDI_SNAPSHOT", "VDI_RESIZE", "VDI_MIRROR",
+                "VDI_GENERATE_CONFIG",
+                "VDI_RESET_ON_BOOT/2", "ATOMIC_PAUSE", "VDI_CONFIG_CBT",
+                "VDI_ACTIVATE", "VDI_DEACTIVATE", "THIN_PROVISIONING", "VDI_READ_CACHING"]
+
+CONFIGURATION = [['device', 'local device path (required) (e.g. /dev/sda3)']]
+
+DRIVER_INFO = {
+    'name': 'Large Block SR',
+    'description': 'SR plugin which emulates a 512 bytes disk on top of a 4KiB device then create a EXT SR',
+    'vendor': 'Vates',
+    'copyright': '(C) 2024 Vates',
+    'driver_version': '1.0',
+    'required_api_version': '1.0',
+    'capabilities': CAPABILITIES,
+    'configuration': CONFIGURATION
+}
+
+LARGEBLOCK_PREFIX = "XSLocalLargeBlock-"
+
+class LargeBlockSR(EXTSR.EXTSR):
+    """Emulating 512b drives for EXT storage repository"""
+
+    DRIVER_TYPE = "largeblock"
+    LOOP_SECTOR_SIZE = 512
+
+    @staticmethod
+    def handles(srtype):
+        return srtype == LargeBlockSR.DRIVER_TYPE
+
+    def load(self, sr_uuid):
+        super(LargeBlockSR, self).load(sr_uuid)
+        self.is_deleting = False
+        self.vgname = LARGEBLOCK_PREFIX + sr_uuid
+        self.remotepath = os.path.join("/dev", self.vgname, sr_uuid)
+
+    def attach(self, sr_uuid):
+        if not self.is_deleting:
+            self._create_emulated_device()
+            if not self._is_vg_connection_correct(): # Check if we need to redo the connection by parsing `vgs -o vg_name,devices self.vgname`
+                self._redo_vg_connection() # Call redo VG connection to connect it correctly to the loop device instead of the real 4KiB block device
+        super(LargeBlockSR, self).attach(sr_uuid)
+
+    def detach(self, sr_uuid):
+        super(LargeBlockSR, self).detach(sr_uuid)
+        if not self.is_deleting:
+            self._destroy_emulated_device()
+
+    @deviceCheck
+    def create(self, sr_uuid, size):
+        base_devices = self.dconf["device"].split(",")
+        for dev in base_devices:
+            logical_blocksize = util.pread2(["blockdev", "--getss", dev])
+            if int(logical_blocksize) == 512:
+                raise xs_errors.XenError("LargeBlockIncorrectBlocksize", opterr="The logical blocksize of the device {} isn't incompatible with other SR types".format(dev))
```

Comment:
`with other SR types`?

Reply:
What would you like as message here? It's supposed to say that users can use any other SR type, e.g. EXTSR.
We could say they should use EXTSR directly maybe?

Reply:
"isn't incompatible"? Is the double negation wanted? if so a rephrasing would be clearer